src/server.ts

```diff
@@ -103,6 +105,27 @@ class Honeybadger extends Client {
       req.end()
     })
   }
+
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  public run<R>(handler: (...args: never[]) => R, onError?: (...args: any[]) => any): R|void {
+    const storeObject = this.__getStoreOrDefaultObject();
+    this.__setStore(AsyncStore)
+
+    if (onError) {
+      // ALS is fine for context-tracking, but `domain` allows us to catch errors
+      // thrown asynchronously (timers, event emitters)
+      // We can't use unhandledRejection/uncaughtException listeners; they're global and shared across all requests
+      // But the `onError` handler might be request-specific.
+      // Note that this doesn't still handle all cases. `domain` has its own problems:
+      // See https://github.com/honeybadger-io/honeybadger-js/pull/711
+      const dom = domain.create();
+      const onErrorWithContext = (err) => this.__store.run(storeObject, () => onError(err));
```

Comment:
Doesn't this create a new context? I think that every time we call `ALS.run()` it creates a new isolated context. So if something was added to the store (in the `handler`) before throwing the error event is raised, we will not see it here because the `storeObject` is the initial object we retrieved at the beginning (before calling the `handler`).
Do you agree? Can we write a test for this?
Maybe it would be better if we wrapped all of this inside `this.store.run()` ? Like:
```
const storeObject = this.__getStoreOrDefaultObject();
this.__setStore(AsyncStore)

return this.__store.run(storeObject, () => {
  if (onError) {
    // ALS is fine for context-tracking, but `domain` allows us to catch errors
    // thrown asynchronously (timers, event emitters)
    // We can't use unhandledRejection/uncaughtException listeners; they're global and shared across all requests
    // But the `onError` handler might be request-specific.
    // Note that this doesn't still handle all cases. `domain` has its own problems:
    // See https://github.com/honeybadger-io/honeybadger-js/pull/711
    const dom = domain.create();
    dom.on('error', onError);
    handler = dom.bind(handler);
  }

  return handler();
});
```

Reply:
Hehe, I went through the same thought patterns and iterations too.😅

Your suggestion doesn't work because it seems the domain error handler is called *outside* the context of the ALS context (even though we ran the domain inside it). A simple script you can test with:

```js
const als = new (require("async_hooks").AsyncLocalStorage);

als.run({a: true}, () => {
    const dom = require("domain").create();
    onError = () => console.log(als.getStore())
    dom.on('error', onError);
    handler = dom.bind(() => { throw new Error });
    handler();
});
```

Logs `undefined`.

So apparently, the ALS context ends before the domain error handler is called.

As for your first question, the context is an object, so it is preserved, since we're passing the same object to both cases of `als.run`. Test:

```js
const als = new (require("async_hooks").AsyncLocalStorage);

ctx = {a: true}
als.run(ctx, () => {
    const dom = require("domain").create();
    onError = () => console.log(als.getStore())
    dom.on('error', (e) => als.run(ctx, onError));
    handler = dom.bind(() => { 
        als.getStore().b = true;
        throw new Error
    });
    handler();
});
```

Logs:

```
{ a: true, b: true }
```

Reply:
Also, re tests: the `captures errors in timers with the right context` below tests for this. The context object is set in the handler, then we grab it in the error handler and assert that it's correct.

Reply:
Thorough explanation, thanks!